### ISSUE TITLE
GVT-g VM crashes and freezes the host

### ISSUE THREAD

**Issue by dgarfias** [01org/gvt-linux#188]
So I had been using GVT-g for over a year with no problems but since last month my Windows 10 VM crashes (and it takes the host with it) when I launch any app that is more demanding than a web browser or a PDF viewer. For example, this now always happens when I launch AutoCAD Electrical, but it didn't happen before.

Here is the output of journalctl -k https://pastebin.com/raw/hEg17TdG but I can't find any error messages that would explain the crash.

I was wondering if anyone has the same problem. Some information about my system:

CPU: Intel i5-10210U, iGPU: Intel UHD (CML GT2)

OS: Arch Linux, DE: GNOME 40, Kernel: 5.12.9-arch1-1, QEMU version: 6.0.0

**Comment by sochotnicky** [comment]
I started seeing something similar recently (or identical - I'll leave the devs to ponder that based on below info).

I believe it started in 5.12.x kernel (i.e. 5.11.x was working fine). I run a Windows 10 VM and it's basically guaranteed to hard-freeze the host kernel to the point of not even sys-rq can recover/reboot. As such I don't have any logs available but I am more than happy to try and gather some info.

In my cases the used components are:
CPU: i7-9700K
OS: Gentoo, any kernel in 5.12.x line (currently can still reproduce in 5.12.11), QEMU 6.0.0 but I could also reproduce on older 5.2.0.

**Comment by dgarfias** [comment]
> I started seeing something similar recently (or identical - I'll leave the devs to ponder that based on below info).
> 
> I believe it started in 5.12.x kernel (i.e. 5.11.x was working fine). I run a Windows 10 VM and it's basically guaranteed to hard-freeze the host kernel to the point of not even sys-rq can recover/reboot. As such I don't have any logs available but I am more than happy to try and gather some info.
> 
> In my cases the used components are:
> CPU: i7-9700K
> OS: Gentoo, any kernel in 5.12.x line (currently can still reproduce in 5.12.11), QEMU 6.0.0 but I could also reproduce on older 5.2.0.

Okay, in that case I will try a kernel downgrade to confirm if we have the same problem.

**Comment by TerrenceXu** [comment]
The regression is introduced by kernel 5.12, you can use the 5.11, it is stable and validated by us.

**Comment by sochotnicky** [comment]
Ah, well good to hear that it's a known problem in 5.12 and I wasn't
imagining things. Guess we can expect for it to be sorted in 5.13 once it
comes out ?

**Comment by dgarfias** [comment]
I downgraded the kernel to version 5.11 and can confirm that the issue is no longer present.

**Comment by mriedel87** [comment]
Hi guys,
it's also crashing with vanilla kernel 5.13.
Do you have a fix? It would be great to know the pull request :) 

Best regards

Marc

**Comment by dgarfias** [comment]
Any news on this issue? What's causing it? I don't want to be stuck with an older kernel.

**Comment by TerrenceXu** [comment]
@dgarfias , we are working on the bug fix, the regression commit as below, you can revert it in any kernel you want to use first. :)

commit 58586680ffadc37636120d9f59955aa5f7a32b7d
Author: Chris Wilson <chris@chris-wilson.co.uk>
Date: Mon Jan 25 22:01:52 2021 +0000
 
drm/i915: Disable atomics in L3 for gen9

**Comment by dgarfias** [comment]
> @dgarfias , we are working on the bug fix, the regression commit as below, you can revert it in any kernel you want to use first. :)
> 
> commit [5858668](https://github.com/intel/gvt-linux/commit/58586680ffadc37636120d9f59955aa5f7a32b7d)
> Author: Chris Wilson [chris@chris-wilson.co.uk](mailto:chris@chris-wilson.co.uk)
> Date: Mon Jan 25 22:01:52 2021 +0000
> 
> drm/i915: Disable atomics in L3 for gen9

Thanks, I'll try it :)

**Comment by SimpliFly03** [comment]
> @dgarfias , we are working on the bug fix, the regression commit as below, you can revert it in any kernel you want to use first. :)
> 
> commit [5858668](https://github.com/intel/gvt-linux/commit/58586680ffadc37636120d9f59955aa5f7a32b7d)
> Author: Chris Wilson [chris@chris-wilson.co.uk](mailto:chris@chris-wilson.co.uk)
> Date: Mon Jan 25 22:01:52 2021 +0000
> 
> drm/i915: Disable atomics in L3 for gen9

I also had similar problem though not exactly the same. In my case while the Windows guest act broken and display is glitched and frozen, the host only have some momentary freezes every 10-15 seconds. If I kill the VM or shut down through ssh, the host returns to proper functioning as if none of those glitches or freezes happened. Of course starting the VM again problem returns. Around that time I was very busy so I didn't have enough time to debug and report it, and Google wasn't helpful either. Thus, I have just booted to 5.11 kernel every time I need VM with GVT-G.

But today I seen your comment. I applied that commit as a revert to 5.13.1 kernel and tested it. And result is no glitch or freeze anymore and the VM is functioning properly like back in 5.11. The host is Arch Linux, the guest is Windows 11. 

Thanks for offering solution.

**Comment by mriedel87** [comment]
Hi guys,

even with the latest vanilla kernel 5.14 (and also 5.13.13), which includes the mentioned fix "drm/i915/gvt: Fix cached atomics setting for Windows VM", it's still crashing.

Aug 30 10:20:37 kaby kernel: [  233.546504] i915 0000:00:02.0: [drm] Resetting rcs0 for preemption time out
Aug 30 10:20:37 kaby kernel: [  233.547026] i915 0000:00:02.0: [drm] GPU HANG: ecode 9:1:8ed9fff2
Aug 30 10:20:48 kaby kernel: [  244.040421] Asynchronous wait on fence 0000:00:02.0:kwin_x11[2654]:2482 timed out (hint:intel_atomic_commit_ready [i915])
Aug 30 10:20:51 kaby kernel: [  247.560738] i915 0000:00:02.0: [drm] GPU HANG: ecode 9:0:00000000
Aug 30 10:20:51 kaby kernel: [  247.561744] i915 0000:00:02.0: [drm] Resetting rcs0 for stopped heartbeat on rcs0
Aug 30 10:20:51 kaby kernel: [  247.562276] i915 0000:00:02.0: [drm] Resetting chip for stopped heartbeat on rcs0
Aug 30 10:20:51 kaby kernel: [  247.664819] [drm:__uc_sanitize [i915]] *ERROR* Failed to reset GuC, ret = -110
Aug 30 10:20:51 kaby kernel: [  247.736823] i915 0000:00:02.0: [drm] *ERROR* Failed to reset chip
Aug 30 10:20:51 kaby kernel: [  247.736825] i915 0000:00:02.0: [drm:add_taint_for_CI [i915]] CI tainted:0x9 by intel_gt_reset+0x2c8/0x2f0 [i915]

Best regards,

Marc

**Comment by SimpliFly03** [comment]
@mriedel87 it is strange I am currently running 5.14 kernel and GVT-G on Windows 11 VM works just fine. Maybe our problems were different and what they fixed was mine. Also, recently, an Intel driver update inside of VM broken GVT-G and causes Qemu(not Windows) to instantly crash a few seconds after the GPU driver was initialized. I fixed it by booting into the safe mode and downgrading the driver. Currently running the driver 28.20.100.8322 and it works fine. 

Edit: I have just switched to driver 30.0.100.9805 it also works just fine.

**Comment by mriedel87** [comment]
Hi,

I've updated my Windows 10 VM to the latest Intel DCH driver, but it still crashes. When I set i915.enable_guc to 0 instead of 2, I had no freeze so far. But I'd like to mention, that i915.enable_guc=2 worked with Kernel 5.11 - maybe there lies the regression?

I'd also like to know, if your bare metal boot time significantly decreases when you disable GVT? When enabled, the kernel seems to add 60 seconds for the initialization of the i915 driver.

Best regards,

Marc

**Comment by SimpliFly03** [comment]
> Hi,
> 
> I've updated my Windows 10 VM to the latest Intel DCH driver, but it still crashes. When I set i915.enable_guc to 0 instead of 2, I had no freeze so far. But I'd like to mention, that i915.enable_guc=2 worked with Kernel 5.11 - maybe there lies the regression?
> 
> I'd also like to know, if your bare metal boot time significantly decreases when you disable GVT? When enabled, the kernel seems to add 60 seconds for the initialization of the i915 driver.
> 
> Best regards,
> 
> Marc

According to [Arch Wiki](https://wiki.archlinux.org/title/Intel_GVT-g#Prerequisite) this is already what you are suppose to do. I don't know how it was worked back in 5.11.

Also I don't have any boot time problems regarding GVTg. Maybe you should try [early KMS](https://wiki.archlinux.org/title/kernel_mode_setting#Early_KMS_start).

**Comment by mriedel87** [comment]
Regarding early KMS,

the following files are included in the initramfs:

usr/lib/firmware/i915
usr/lib/firmware/i915/adls_dmc_ver2_01.bin
usr/lib/firmware/i915/bxt_dmc_ver1_07.bin
usr/lib/firmware/i915/bxt_guc_49.0.1.bin
usr/lib/firmware/i915/bxt_huc_2.0.0.bin
usr/lib/firmware/i915/cml_guc_49.0.1.bin
usr/lib/firmware/i915/cml_huc_4.0.0.bin
usr/lib/firmware/i915/cnl_dmc_ver1_07.bin
usr/lib/firmware/i915/dg1_dmc_ver2_02.bin
usr/lib/firmware/i915/ehl_guc_49.0.1.bin
usr/lib/firmware/i915/ehl_huc_9.0.0.bin
usr/lib/firmware/i915/glk_dmc_ver1_04.bin
usr/lib/firmware/i915/glk_guc_49.0.1.bin
usr/lib/firmware/i915/glk_huc_4.0.0.bin
usr/lib/firmware/i915/icl_dmc_ver1_09.bin
usr/lib/firmware/i915/icl_guc_49.0.1.bin
usr/lib/firmware/i915/icl_huc_9.0.0.bin
usr/lib/firmware/i915/kbl_dmc_ver1_04.bin
usr/lib/firmware/i915/kbl_guc_49.0.1.bin
usr/lib/firmware/i915/kbl_huc_4.0.0.bin
usr/lib/firmware/i915/rkl_dmc_ver2_02.bin
usr/lib/firmware/i915/skl_dmc_ver1_27.bin
usr/lib/firmware/i915/skl_guc_49.0.1.bin
usr/lib/firmware/i915/skl_huc_2.0.0.bin
usr/lib/firmware/i915/tgl_dmc_ver2_08.bin
usr/lib/firmware/i915/tgl_guc_49.0.1.bin
usr/lib/firmware/i915/tgl_huc_7.5.0.bin
usr/lib/modules/5.14.0-2.slh.1-aptosid-amd64/kernel/drivers/gpu/drm/i915
usr/lib/modules/5.14.0-2.slh.1-aptosid-amd64/kernel/drivers/gpu/drm/i915/gvt
usr/lib/modules/5.14.0-2.slh.1-aptosid-amd64/kernel/drivers/gpu/drm/i915/gvt/kvmgt.ko
usr/lib/modules/5.14.0-2.slh.1-aptosid-amd64/kernel/drivers/gpu/drm/i915/i915.ko

The module intel_agp is of no concern for my system.

**Comment by SimpliFly03** [comment]
> The module intel_agp is of no concern for my system.

I don't have that either in my initramfs. Also, could you try to create a file `/etc/modprobe.d/i915.conf` and include it in your initramfs with the contents below:

```
options i915 enable_guc=0 enable_gvt=1 enable_fbc=0 modeset=1
```

**Comment by mriedel87** [comment]
I added the options via kernel cmd and checked the option values in /sys/module/i915/parameters.
Sadly nothing changed. There is always a delay of 60 seconds.

**Comment by thalesbarretto01** [comment]
Hello there.
I've been facing the same issue and somehow managed to make gvt-g work under 5.14.

During a sysupgrade i've notice the module "vfio_mdev" was missing for 5.14 .
In previous versions, the module vfio_mdev was included.
But the 5.14 module name is just "mdev" (maybe consolidates the old vfio_mdev).

By  changing the mkinitcpio.conf modules list and regenerating initramfs made my VM's work again.

e.g.
up to 5.11
MODULES=( kvmgt vfio-iommu-type1 vfio-mdev) 

changed to (working in 5.14 and 5.10 - lts)
MODULES=( kvmgt vfio-iommu-type1 mdev vfio-mdev vfio_mdev)

The, run mkinitcpio -P and reboot

Please give it a try

**Comment by mriedel87** [comment]
@thalesbarretto01
Hi, to what problem are you referring?
According to this [commit](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?h=v5.15-rc2&id=af3ab3f9b986cdbc1b97b8a3341ce78851edb0dd), vfio-mdev was removed and only mdev is required (which is included in my initramfs).

**Comment by thalesbarretto01** [comment]
> @thalesbarretto01
> Hi, to what problem are you referring?
> According to this [commit](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?h=v5.15-rc2&id=af3ab3f9b986cdbc1b97b8a3341ce78851edb0dd), vfio-mdev was removed and only mdev is required (which is included in my initramfs).

@mriedel87.
Hi, from kernel 12. till 5.14 the VM's using vfio-mdev crashed when launching, with a journal entry "CPUHANG".
Yes,  i've noticed the change you mentioned about "vfio-mdev" module.
But, after some testing, with the new 5.14.6-arch1-1 kernel  the VM's using GVT-g things apparently started working again.

**Comment by h3x4d3c1m4l** [comment]
It crashes at my system (Arch Linux with 5.14.7 kernel) with exactly these very same lines of kernel log:

```
i915 0000:00:02.0: [drm] Resetting rcs0 for preemption time out
i915 0000:00:02.0: [drm] GPU HANG: ecode 9:1:8ed9fff2
```
But it stops immediately after that. There is no reset or anything. The whole system freezes and also stops being reachable over the network.

I haven't found a workaround yet. Perhaps it has something to do with my Dell XPS15 9500's very low aperture size (256 MB). Sometimes I can create the virtual GPU, sometimes it says something like "no space left"...

Or it might have something to do with me having forcing the resolution to 4k 60 Hz because of my notebook's screen. This is definitely unsupported as the max display setting is listed as 1920x1200 60Hz for my mdev setting... I'm still testing currently.

**Comment by SimpliFly03** [comment]
@h3x4d3c1m4l Is the VM freezes or the whole host. Since I upgraded my kernel to 5.14.6 I started to have host freezes few minutes (<10mins) after VM boot that sometimes occur sometimes doesn't. It is totally random. I know that 5.14.1 works just fine but I don't know anything about versions between 5.14.1 and 5.14.6 since I haven't run these versions. Also, I use a resolution that is under the limits (1920x1080 60Hz).

**Comment by h3x4d3c1m4l** [comment]
While the last crash occured the resolution was at 1920x1080 which should be fine...

It's the whole host that freezes, the two log messages are logged by the host right before it totally freezes. Using `journalctl` I can get the logs from the previous boot but that even doesn't show the `GPU HANG` (it probably crashed too fast for the lines to be actually written to the disk). I captured the log lines by opening a SSH connection from another machine. After those two lines were written to the terminal, the connection freezed and the SSH client quitted with a broken pipe error.

**Comment by rahulc07** [comment]
I can confirm that I have a similar issue happening, but it is only with guests running the linux 5.14 kernel everything else works, I also get a bad address error when doing this.  Also windows has some weird issues with crashing and stuff under seabios I can only get it to work with vbios_gvt_uefi.rom and uefi.  Even reinstalling did not fix it.  Also reverting the kernel did not fix it

Host:
Arch Linux 
Custom Compiled 5.14.7 Kernel 

Guest:
Debian 11 Bullseye with gnome (wayland)
Custom Compiled 5.14.1 Kernel

**Comment by dgarfias** [comment]
> So I had been using GVT-g for over a year with no problems but since last month my Windows 10 VM crashes (and it takes the host with it) when I launch any app that is more demanding than a web browser or a PDF viewer. For example, this now always happens when I launch AutoCAD Electrical, but it didn't happen before.
> 
> Here is the output of journalctl -k https://pastebin.com/raw/hEg17TdG but I can't find any error messages that would explain the crash.
> 
> I was wondering if anyone has the same problem. Some information about my system:
> 
> CPU: Intel i5-10210U, iGPU: Intel UHD (CML GT2)
> 
> OS: Arch Linux, DE: GNOME 40, Kernel: 5.12.9-arch1-1, QEMU version: 6.0.0

I'm still having the same problem with kernel 5.14.9

**Comment by juliotux** [comment]
Confirmed in kernel 5.14.11, ArchLinux. i7-9750H and QEMU 6.1.0.

**Comment by rahulc07** [comment]
Confirmed working in 5.11 with i9-10900 (UHD 630) 5.11 kernel 

(Meaning that GVT-g is working in that kernel but not working in 5.14)

**Comment by patrykk** [comment]
I have tested iommu passthrough with kernel 5.14.14 and 5.15-rc7. With the kernel option iommu.passthrough=1 the gvt windows 10 virtual machine is stable. Please test the option with options i915.enable_fbc=0 i915.enable_guc=0 i915.enable_gvt=1 i915.enable_execlists=0. Please remove option iommu.strict or set it to 0. Best regards.

**Comment by juliotux** [comment]
> I have tested iommu passthrough with kernel 5.14.14 and 5.15-rc7. With the kernel option iommu.passthrough=1 the gvt windows 10 virtual machine is stable. Please test the option with options i915.enable_fbc=0 i915.enable_guc=0 i915.enable_gvt=1 i915.enable_execlists=0. Please remove option iommu.strict or set it to 0. Best regards.

Hi @patrykk . The only option in this list I didn't test is the execlist. This problem is very random. Some times it takes hours to freeze the host. Some times it freezes in just minutes. I could not find too any specific reason that caused the crash. I will try this option and send a feedback.

**Comment by patrykk** [comment]
Hi. Here you can find my whole menuentry from grub for default arch linux kernel:

menuentry "arch linux in 1st partition of /dev/sda gpt partition table" {
insmod lvm
set root=(lvm/ssd2-arch)
linux /boot/vmlinuz-linux root=/dev/mapper/ssd2-arch ro resume=/dev/mapper/ssd2-swap intel_iommu=on kvm.ignore_msrs=1 acpi_enforce_resources=lax loglevel=3 init=/usr/lib/systemd/systemd fastboot spectre_v2=retpoline     spectre_v2_user=off spec_store_bypass_disable=on i915.enable_fbc=0 i915.fastboot=1 pcie_port_pm=on noibrs noibpb l1tf=off no_stf_barrier mds=off rd.luks.options=discard  i915.enable_guc=0 i915.enable_gvt=1 i915.modeset=1 i915.enable_execlists=0 iommu.passthrough=1
 initrd /boot/initramfs-linux.img

}

Configuration for hugepahes:
cat /sys/kernel/mm/transparent_hugepage/enabled 
always [**madvise**] never

and sysctl:
sysctl -p
net.ipv4.tcp_syncookies = 1
net.ipv4.ip_forward = 0
net.ipv6.conf.all.forwarding = 0
vm.vfs_cache_pressure = 100
vm.laptop_mode = 1
vm.swappiness = 0
fs.inotify.max_user_watches = 524288
kernel.sysrq = 1

**Comment by pongo1231** [comment]
Can confirm `iommu.passthrough=1` fixed the host system lock-ups for me (at least from my testing, as mentioned the issue is somewhat random and it could've just been a coincidence, though I rebooted the host each time between my tests to make sure). Interestingly enough I could barely find any kind of information about that kernel parameter so I'm not sure what its purpose is specifically.

Arch Linux with the 5.14.15-zen1-1-zen kernel and the following i915 module parameters: `enable_guc=0 enable_fbc=1 fastboot=1 enable_gvt=1`

**Comment by Maryse47** [comment]
In nutshell this option weakens iommu security and increase performance. Technical details can be found [here](https://lenovopress.com/lp1467.pdf)

**Comment by patrykk** [comment]
> In nutshell this option weakens iommu security and increase performance. Technical details can be found [here](https://lenovopress.com/lp1467.pdf)

Yes. That's right and I have known about it. I dont want to use it because performance difference is only around 5% but with option passthrough gvt is stable for me.

**Comment by h3x4d3c1m4l** [comment]
I have always used `iommu=pt` which should be (according to the [document](https://lenovopress.com/lp1467.pdf) that @Maryse47 just wrote about) the same as `iommu.passthrough=1`. Yet it's still unstable for me (random host freezes).

But I'm testing the stable 5.15 mainline kernel right now. If that fails, I will replace the `iommu=pt` by `iommu.passthrough=1` and test again. If that fails too I will test with neither option in the cmdline...

**Comment by h3x4d3c1m4l** [comment]
Host and VM now both have 8 hrs of uptime, still no crash...

Kernel cmdline:
`i915.enable_gvt=1 i915.enable_fbc=1 intel_iommu=on iommu=pt`

Output of `uname -a`:
`Linux h3x-laptop 5.15.0-1-mainline #1 SMP PREEMPT Mon, 01 Nov 2021 06:34:19 +0000 x86_64 GNU/Linux`

EDIT: It crashed anyway, first without crashing the host, then I rebooted the VM and it kept hanging on the Windows 10 bootloader spinner, then I forced rebooted the VM. It then booted fine, but crashed the host after about 15 minutes. Now testing with the `iommu.passthrough=1` option instead of `iommu=pt`.

EDIT 2: It crashed again... gonna try ASAP without any of the two options.

EDIT 3: Still unstable :-1: I'm out of options at the moment...

**Comment by patrykk** [comment]
Hi. I have update for my configuration. I have tested that same configuration on the second machine and the windows 10 guest hangs randomly even with iommu.passthrough=1 . I have realized that the problem occure when guest support  screen power management. So now I reproduce stable environment when on the host I add iommu.passthrough=1 and configure guest like below:
powercfg /L      # list power shemes
powercfg /S  8c5e7fda-e8bf-4a96-9a85-a6e23a8c635c          # ( set performance scheme )
powercfg /hibernate off
powercfg -change monitor-timeout-ac 0 
powercfg -change monitor-timeout-dc 0

**Comment by dgarfias** [comment]
> Hi. I have update for my configuration. I have tested that same configuration on the second machine and the windows 10 guest hangs randomly even with iommu.passthrough=1 . I have realized that the problem occure when guest support screen power management. So now I reproduce stable environment when on the host I add iommu.passthrough=1 and configure guest like below: powercfg /L # list power shemes powercfg /S 8c5e7fda-e8bf-4a96-9a85-a6e23a8c635c # ( set performance scheme ) powercfg /hibernate off powercfg -change monitor-timeout-ac 0 powercfg -change monitor-timeout-dc 0

Have you tried running graphics intensive tasks? What kernel version are you running? This issue I'm having is not only a guest freeze, but also a host freeze, and it occurs mostly when running graphic intensive tasks. It only occurs with kernel versions 5.12+

I had also the power management issue long before but I solved it disabling power management, but that only caused a guest hang.

**Comment by patrykk** [comment]
Yes. That issue freeze guest and kills qemu process. Host couldn't release memory and We need to perform reboot. I performed tests with unigine heaven, unigine valley, video streaming and game like Universal sanbox. I have executed all programs  in the same time. 
Tested with kernels 5.14.14, 5.14.15, 5.15 and 5.15.1.

Power management problem lead to hangs also qemu which lead process to notinteruptable stage with huge kernel log.
When I configure both, host and guest we have stable environment for a week.
btw. one virtual machine with gvt-g per host. 

config file: https://pastebin.com/4qdgsZpP

**Comment by h3x4d3c1m4l** [comment]
@patrykk Your issue might not be related to this issue, as your kernel does not seem to completely freeze. I would advise to create another issue to explain the situation like you did in your last message.

I did test monitor DPMS using this Powershell command I found on StackOverflow (which should put the monitors in sleep programmatically so you don't have to wait for the Windows Power Management), but I can't reproduce it:
```(Add-Type -MemberDefinition "[DllImport(""user32.dll"")]`npublic static extern int SendMessage(int hWnd, int hMsg, int wParam, int lParam);" -Name "Win32SendMessage" -Namespace Win32Functions -PassThru)::SendMessage(0xffff, 0x0112, 0xF170, 2)```

The guest screen does freeze, but when I press a button or move the mouse it continues to work just fine.

**Comment by pongo1231** [comment]
The issue re-surfaced for me now after a few days, but in a different form: while opening applications works fine (including ones that demand the GPU), if I leave my win10 VM running for long enough (~5 mins) without any kind of user input it'll lock up the host to a point where not even sysrq works. The power management config mentioned above unfortunately didn't help. Any idea how I could fetch logs from that state? Trying to ssh in from another machine doesn't work.

**Comment by h3x4d3c1m4l** [comment]
@pongo1231 Maybe try a serial connection... if it's a laptop (of desktop without native serial port) maybe try a USB to serial bridge...

**Comment by rugubara** [comment]
> The issue re-surfaced for me now after a few days, but in a different form: while opening applications works fine (including ones that demand the GPU), if I leave my win10 VM running for long enough (~5 mins) without any kind of user input it'll lock up the host to a point where not even sysrq works. The power management config mentioned above unfortunately didn't help. Any idea how I could fetch logs from that state? Trying to ssh in from another machine doesn't work.

I use EFI_VARS pstore to capture the last breath of the kernel. https://cateee.net/lkddb/web-lkddb/EFI_VARS_PSTORE.html 

Once you have enabled it, keep an eye on /sys/fs/pstore - and clean files from there on a regular basis.

**Comment by pongo1231** [comment]
> > The issue re-surfaced for me now after a few days, but in a different form: while opening applications works fine (including ones that demand the GPU), if I leave my win10 VM running for long enough (~5 mins) without any kind of user input it'll lock up the host to a point where not even sysrq works. The power management config mentioned above unfortunately didn't help. Any idea how I could fetch logs from that state? Trying to ssh in from another machine doesn't work.
> 
> I use EFI_VARS pstore to capture the last breath of the kernel. https://cateee.net/lkddb/web-lkddb/EFI_VARS_PSTORE.html
> 
> Once you have enabled it, keep an eye on /sys/fs/pstore - and clean files from there on a regular basis.

Thanks! Seems like the issue doesn't occur for me anymore, not sure why. I'll post logs here if it happens again.

**Comment by pongo1231** [comment]
While I don't have any issues with the guest causing the host to hang anymore, the guest still occasionally hangs up and I wasn't even able to kill qemu a couple of times (seems like it's causing an exception in the `kvmgt` kernel module? Not sure). However after applying 2 kernel patches it's been completely stable and I haven't had the guest freeze on me once in a week with frequent usage. This is on a 5.15.5 kernel with an i5-9300H CPU.

First off I tried to revert the commit mentioned [here](https://github.com/intel/gvt-linux/issues/188#issuecomment-877300477). This however didn't work as that commit also contained some other changes and reverting it would cause an error while compiling. I've instead created a [patch](https://gist.github.com/pongo1231/3df38a08ab601350050ea280eb83e9a2) which reverts only the relevant part, which worked fine.

Also there's this [patch](https://github.com/intel/gvt-linux/issues/153#issuecomment-858081285) which I also saw mentioned and which applied without any issues. I've applied both of these patches at the same time so I'm not sure if only one of them is actually necessary to resolve this issue in this case.

Would be curious if this also resolves any issues with guest / host lock-ups for anyone else?

**Comment by juliotux** [comment]
> While I don't have any issues with the guest causing the host to hang anymore, I'd still have the guest hang and be unable to kill qemu occasionally (seems like it's causing an exception in the `kvmgt` kernel module? Not sure). However after applying 2 kernel patches it's been completely stable and I haven't had the guest freeze on me once in a week with frequent usage. This is on a 5.15.5 kernel with an i5-9300H CPU.
> 
> First off I tried to revert the commit mentioned [here](https://github.com/intel/gvt-linux/issues/188#issuecomment-877300477). This however didn't work as that commit also contained some other changes and reverting it would cause an error while compiling. I've instead created a [patch](https://gist.github.com/pongo1231/3df38a08ab601350050ea280eb83e9a2) which reverts only the relevant part, which worked fine.
> 
> Also there's this [patch](https://github.com/intel/gvt-linux/issues/153#issuecomment-858081285) which I also saw mentioned and which applied without any issues. I've applied both of these patches at the same time so I'm not sure if only one of them is actually necessary to resolve this issue in this case.
> 
> Would be curious if this also resolves any issues with guest / host lock-ups for anyone else?

Using the machine all day long with the patches and it is working. For arch users, I crated a custom kernel PKGBUILD to apply the patches. The repository is: [archlinux-package-linux-gvtg](https://github.com/juliotux/archlinux-package-linux-gvtg)

**Comment by ghost** [comment]
> > While I don't have any issues with the guest causing the host to hang anymore, I'd still have the guest hang and be unable to kill qemu occasionally (seems like it's causing an exception in the `kvmgt` kernel module? Not sure). However after applying 2 kernel patches it's been completely stable and I haven't had the guest freeze on me once in a week with frequent usage. This is on a 5.15.5 kernel with an i5-9300H CPU.
> > 
> > First off I tried to revert the commit mentioned [here](https://github.com/intel/gvt-linux/issues/188#issuecomment-877300477). This however didn't work as that commit also contained some other changes and reverting it would cause an error while compiling. I've instead created a [patch](https://gist.github.com/pongo1231/3df38a08ab601350050ea280eb83e9a2) which reverts only the relevant part, which worked fine.
> > 
> > Also there's this [patch](https://github.com/intel/gvt-linux/issues/153#issuecomment-858081285) which I also saw mentioned and which applied without any issues. I've applied both of these patches at the same time so I'm not sure if only one of them is actually necessary to resolve this issue in this case.
> > 
> > Would be curious if this also resolves any issues with guest / host lock-ups for anyone else?
> 
> Using the machine all day long with the patches and it is working. For arch users, I crated a custom kernel PKGBUILD to apply the patches. The repository is: [archlinux-package-linux-gvtg](https://github.com/juliotux/archlinux-package-linux-gvtg)

@juliotux @pongo1231 Seems the patched version also solves the problem for me. Haven't noticed anything wrong on system journal.

**Comment by pg2web** [comment]
Applied two patches discussed here and rebuilt Fedora34 5.15.6-100.fc34.x86_64 kernel from srpm - no more host crashes that I constantly had in Realflight 9.5 simulator running in Win10 LTSC VM. There was one crash (only VM got reboot, no host freeze) when tried to change resolution from OS, but it may be related to spice agent intervention. Thank you!

Just in case attaching combined patch.

[intel-gvt-hostcrash-fix.patch.txt](https://github.com/intel/gvt-linux/files/7701120/intel-gvt-hostcrash-fix.patch.txt)

**Comment by wioo** [comment]
Can someone try @pongo1231 patches on 5.16-arch1-1.
It worked for me with 5.15.* kernels, but with 5.16 qemu freezes (cant kill it and have to forcefully shutdown computer).

**Comment by xmu714** [comment]
> > While I don't have any issues with the guest causing the host to hang anymore, I'd still have the guest hang and be unable to kill qemu occasionally (seems like it's causing an exception in the `kvmgt` kernel module? Not sure). However after applying 2 kernel patches it's been completely stable and I haven't had the guest freeze on me once in a week with frequent usage. This is on a 5.15.5 kernel with an i5-9300H CPU.
> > First off I tried to revert the commit mentioned [here](https://github.com/intel/gvt-linux/issues/188#issuecomment-877300477). This however didn't work as that commit also contained some other changes and reverting it would cause an error while compiling. I've instead created a [patch](https://gist.github.com/pongo1231/3df38a08ab601350050ea280eb83e9a2) which reverts only the relevant part, which worked fine.
> > Also there's this [patch](https://github.com/intel/gvt-linux/issues/153#issuecomment-858081285) which I also saw mentioned and which applied without any issues. I've applied both of these patches at the same time so I'm not sure if only one of them is actually necessary to resolve this issue in this case.
> > Would be curious if this also resolves any issues with guest / host lock-ups for anyone else?
> 
> Using the machine all day long with the patches and it is working. For arch users, I crated a custom kernel PKGBUILD to apply the patches. The repository is: [archlinux-package-linux-gvtg](https://github.com/juliotux/archlinux-package-linux-gvtg)

Your solution worked perfectly! It would be nice to upgrade to 5.16 kernel.

**Comment by juliotux** [comment]
@xmu714 updated!

**Comment by NerdyProjects** [comment]
I also had crashes with the standard kernel and now use the patched archlinux kernels. Is there any chance to see mainline fixes? Is the actual cause clear? How can we help to further debug?

**Comment by TerrenceXu** [comment]
@NerdyProjects 
The issue introduced by this commit https://github.com/intel/gvt-linux/commit/58586680ffadc37636120d9f59955aa5f7a32b7d have not fixed in standard kernel until now.

**Comment by thalesbarretto01** [comment]
There is onde surefire way to trigger the bug on my system (8565u, archlinux kernel since the blamed commit): on a windows guest, start a video call on ms teams, the system will crash (host and guest) and become irresponsive. But i can't even dump, since the the whole system is completely crashed, no trace left. Wish i could help the fix going upstream.@TerrenceXu is there anything we can do to help fixing it?

**Comment by tpressure** [comment]
Please see my coment here: https://github.com/intel/gvt-linux/issues/153#issuecomment-1044747758

**Comment by TerrenceXu** [comment]
The issue gone in the latest gvt-staging tag: [gvt-staging: 2022y-02m-17d-16h-19m-52s](https://github.com/intel/gvt-linux/commit/11bd528bd71f1fd2e97a20e36b05f08602feb506)

**Comment by i1u5** [comment]
> The issue gone in the latest gvt-staging tag: [gvt-staging: 2022y-02m-17d-16h-19m-52s](https://github.com/intel/gvt-linux/commit/11bd528bd71f1fd2e97a20e36b05f08602feb506)

So this can be considered fixed then?

**Comment by TerrenceXu** [comment]
@i1u5 This tag with some i915 refactor work, not for fix this issue individually, we will keep monitor this issue in new tags.

**Comment by ThalesBarretto** [comment]
Testing gvt-staging since [gvt-staging: 2022y-02m-17d-16h-19m-52s](https://github.com/intel/gvt-linux/commit/11bd528bd71f1fd2e97a20e36b05f08602feb506). It's looking good.
For those running archlinux, i've provided [a helping hand](https://github.com/ThalesBarretto/archlinux-intel-gvt-linux) @juliotux

**Comment by NerdyProjects** [comment]
Can we get an update? I don't understand the relevance of the linked gvt-staging tag. Can somebody write in a few easily understandable sentences
a) where the patch currently is in the pipeline
b) an estimation of time until it lands in released kernel?

I'd like to rely on using gvt-g for my daily work but keeping up with the kernel updates on arch seems impossible :-(

Thanks to all involved parties for getting it stable again!

**Comment by wioo** [comment]
I dont know what changed but windows 10 and 11 vms are fine atm. (arch linux)

**Comment by dgarfias** [comment]
Is the last gvt-staging tag already incorporated into the mainline kernel?

**Comment by ThalesBarretto** [comment]
@juliotux @dgarfias @NerdyProjects i've updated the helper to get the appropriate tag, and applied the patches to match it as close as possible to the v5.18.16 kernel. Maybe you can make do with it, since it solves my problem for now (using Qemu6) and i'd like to share it (idk about the upstream/mainline).

**Comment by pg2web** [comment]
> I dont know what changed but windows 10 and 11 vms are fine atm. (arch linux)

Just want to share my most recent experience with GVT-g enabled Win10 LTSC that had constant crashes until I patched the kernel (see my prev. msg). I didn't use this VM for a while and recently updated Fedora 34 kernel to the fresh one (5.17.12.-100) and decided to give it a try without the patch. VM immediately crashed when I started Realflight9 simulator. But also I paid attention that Intel had some updates to HD 630 driver and updated Win10 HD630 driver (at this moment it's 30.0.101.1338 from 1/20/2022). With new windows driver the issue disappeared... so it confirms the fact that sum of two factors play the role in this freeze - the way how the kernel handles it and Win driver that may crash it sending incorrect command. At least with new Win 10 driver it works without issues.

**Comment by dgarfias** [comment]
I tried running my VM with the latest kernel (6.16) and driver release 30.0.101.1122. I have had no issues and there were no crashes, so I think the problem is solved. If anyone else can confirm.

**Comment by wuruiwm** [comment]
pve下  内核Linux 5.15.39-4-pve #1 SMP PVE 5.15.39-4 (Mon, 08 Aug 2022 15:11:15 +0200)
gvt-g运行60天 正常

**Comment by criollojoel10** [comment]
> I tried running my VM with the latest kernel (6.16) and driver release 30.0.101.1122. I have had no issues and there were no crashes, so I think the problem is solved. If anyone else can confirm.

Same problem with [kernel](https://wiki.archlinux.org/title/kernel) vanilla linux.
Problem solved using kernel linux-zen.

**Comment by Mejituu** [comment]
I still have this issue on kernel 6.4.14, The Windows driver seems to be around 27. It comes from June 1, 2021

**Comment by eskalVAR** [comment]
Still same issue. Please see: https://github.com/intel/gvt-linux/issues/153#issuecomment-1866659269

**Comment by etuckeriv** [comment]
I have what seems to be the same issue. I've just installed Windows 10 on the guest VM through virt-manager, got the virtualized igpu passed through, and the system locked up as I was installing the Intel drivers. I happened to have music playing on my host system when it locked up, and the music kept going. I was also able to use my keyboard's play/pause to control the music just fine so I suspected it was just video output that was frozen.

After a couple of hard reboots of my host system, and waiting around for a while to see if it might unfreeze on it's own after Windows was seemingly finished with the driver install (it didn't), and since I had just spun up the VM I decided to delete it and start over, but I ran into the same problem. Again, I heard the sounds of Windows discovering new devices and installing the drivers, but my video output on the host was completely frozen so I couldn't tell what was happening.

I randomly decided to hit "ctrl+alt+del" and then "enter" which successfully logged me out of my host user session and returned video functionality. (I'm running KDE)

When I logged back in, it was a new KDE user session but interestingly enough the VM was still running as it was before I logged out. I opened it and I was actually able to finish the driver install, but when I chose to reboot at the installer prompt it froze again. I continued logging out and back in on my host system and finished the driver install, but video output continues to freeze after a short time of running the vm.

**Installation Details:**

I also followed the [instructions on the Arch Wiki for Intel GVT-G](https://wiki.archlinux.org/title/Intel_GVT-g). After adding the kernel modules and parameters, I did a full install of Windows 10 into a new VM and installed the guest tools from spice-space.org. Then I shut down the VM and made the necessary xml adjustments outlined in the Arch Wiki article to add the vGPU to the guest. After booting with the vGPU, everything is low-resolution, but runs fine until I install the Intel drivers, then the host video output locks.

**HOST Info:**
OS: Debian GNU/Linux trixie/sid x86_64 (Debian 12 testing w/ security updates from unstable)
Host: Latitude E7470 
Kernel: 6.6.8-2-liquorix-amd64 
Shell: bash 5.2.21 
Display Protocol: Wayland 
Resolution: 1920x1080 
DE: Plasma 5.27.9 
WM: kwin 
CPU: Intel i7-6600U (4) @ 2.801GHz 
GPU: Intel Skylake GT2 [HD Graphics 520] 
Memory: 3837MiB / 31873MiB

Libvirt Version: 9.10.0
QEMU Version: 8.1.2 (Debian 1:8.1.2+ds-1)
Virt-Manager Version: 4.1.0

**GUEST VM Info:**
OS: Windows 10 22H2
Chipset: Q35
Firmware: UEFI (using vbios patch, rom file sitting in /var/lib/libvirt)
VCPU: host-passthrough with 2 threads allocated
Memory: 16GB
Disk: 50GB VirtIO
NIC: VirtIO
Display: Spice Server, "None" listen type, OpenGL enabled
MDEV Device vGPU Passed Through (with video device marked as "none"
Intel Driver Version 15.45

I'm going to try with x11 and then with the stock Debian kernel to see if there is any change and will update the thread.

**Comment by etuckeriv** [comment]
> I have what seems to be the same issue. I've just installed Windows 10 on the guest VM through virt-manager, got the virtualized igpu passed through, and the system locked up as I was installing the Intel drivers. I happened to have music playing on my host system when it locked up, and the music kept going. I was also able to use my keyboard's play/pause to control the music just fine so I suspected it was just video output that was frozen.
> 
> After a couple of hard reboots of my host system, and waiting around for a while to see if it might unfreeze on it's own after Windows was seemingly finished with the driver install (it didn't), and since I had just spun up the VM I decided to delete it and start over, but I ran into the same problem. Again, I heard the sounds of Windows discovering new devices and installing the drivers, but my video output on the host was completely frozen so I couldn't tell what was happening.
> 
> I randomly decided to hit "ctrl+alt+del" and then "enter" which successfully logged me out of my host user session and returned video functionality. (I'm running KDE)
> 
> When I logged back in, it was a new KDE user session but interestingly enough the VM was still running as it was before I logged out. I opened it and I was actually able to finish the driver install, but when I chose to reboot at the installer prompt it froze again. I continued logging out and back in on my host system and finished the driver install, but video output continues to freeze after a short time of running the vm.
> 
> **Installation Details:**
> 
> I also followed the [instructions on the Arch Wiki for Intel GVT-G](https://wiki.archlinux.org/title/Intel_GVT-g). After adding the kernel modules and parameters, I did a full install of Windows 10 into a new VM and installed the guest tools from spice-space.org. Then I shut down the VM and made the necessary xml adjustments outlined in the Arch Wiki article to add the vGPU to the guest. After booting with the vGPU, everything is low-resolution, but runs fine until I install the Intel drivers, then the host video output locks.
> 
> **HOST Info:** OS: Debian GNU/Linux trixie/sid x86_64 (Debian 12 testing w/ security updates from unstable) Host: Latitude E7470 Kernel: 6.6.8-2-liquorix-amd64 Shell: bash 5.2.21 Display Protocol: Wayland Resolution: 1920x1080 DE: Plasma 5.27.9 WM: kwin CPU: Intel i7-6600U (4) @ 2.801GHz GPU: Intel Skylake GT2 [HD Graphics 520] Memory: 3837MiB / 31873MiB
> 
> Libvirt Version: 9.10.0 QEMU Version: 8.1.2 (Debian 1:8.1.2+ds-1) Virt-Manager Version: 4.1.0
> 
> **GUEST VM Info:** OS: Windows 10 22H2 Chipset: Q35 Firmware: UEFI (using vbios patch, rom file sitting in /var/lib/libvirt) VCPU: host-passthrough with 2 threads allocated Memory: 16GB Disk: 50GB VirtIO NIC: VirtIO Display: Spice Server, "None" listen type, OpenGL enabled MDEV Device vGPU Passed Through (with video device marked as "none"
> 
> I'm going to try with xorg and then with the stock Debian kernel to see if there is any change and will update the thread.

Okay, this seems to be related to Wayland, at least in my case. I logged out of my Wayland session and back in with x11 and it seems to be working perfectly.

**Comment by etuckeriv** [comment]
> I'm going to try with x11 and then with the stock Debian kernel to see if there is any change and will update the thread.

Okay, this seems to be related to Wayland, at least in my case. I logged out of my Wayland session and back in with x11 and it seems to be working perfectly.

**UPDATE:** I also rebooted to the stock Debian kernel and tried both Wayland and x11. Confirmed that it still occurs on Wayland with stock kernel and NOT on x11 with stock kernel. So for me this issue definitely seems to be related to Wayland, regardless of kernel.

**Comment by etuckeriv** [comment]
I just realized I had been running this setup before on Gnome with no problems, so I fired up Gnome on a Wayland session and it doesn't have the issue. So for me it seems to be specifically related to KDE on Wayland.

**Comment by pongo1231** [comment]
Can confirm running an igvt-g VM under KDE Plasma on Wayland causes the compositor to lock up. This should probably be reported to KDE's bug tracker for kwin. I haven't tried the Plasma 6 beta to see if it still occurs there.

**Comment by pongo1231** [comment]
Got around to testing it with plasma 6 (beta 2) and running a gvt-g VM does not cause the compositor to lock up unlike plasma 5. Can anyone else replicate this?

**Comment by lyc8503** [comment]
Windows 11 GVT-g VM with latest Intel driver, Host `Linux pve 6.5.11-7-pve #1 SMP PREEMPT_DYNAMIC PMX 6.5.11-7 (2023-12-05T09:44Z) x86_64 GNU/Linux`

The QEMU itself hangs randomly after days of power on.
